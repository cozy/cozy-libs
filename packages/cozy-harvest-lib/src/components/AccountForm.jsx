import React, { PureComponent } from 'react'
import { Form, Field as FinalFormField } from 'react-final-form'
import PropTypes from 'react-proptypes'

import Button from 'cozy-ui/react/Button'
import { translate, extend } from 'cozy-ui/react/I18n'
import Field from 'cozy-ui/react/Field'

import AccountFormError from './AccountFormError'
import {
  getEncryptedFieldName,
  getFieldPlaceholder,
  sanitizeSelectProps
} from '../helpers/fields'
import { KonnectorJobError } from '../helpers/konnectors'
import manifest from 'helpers/manifest'
import OAuthForm from './OAuthForm'

const predefinedLabels = [
  'answer',
  'birthdate',
  'code',
  'date',
  'email',
  'firstname',
  'lastname',
  'login',
  'password',
  'phone'
]

// Out of scope labels already used, should be transferred directly in manifests
// in the future.
const legacyLabels = [
  'branchName' // Used in banking konnectors
]

const VALIDATION_ERROR_REQUIRED_FIELD = 'VALIDATION_ERROR_REQUIRED_FIELD'

export class AccountField extends PureComponent {
  render() {
    const {
      disabled,
      hasError,
      initialValue,
      label,
      name,
      onKeyUp,
      required,
      role,
      t,
      type
    } = this.props

    // Allow manifest to specify predefined label
    const localeKey =
      predefinedLabels.includes(label) || legacyLabels.includes(label)
        ? label
        : name

    const isEditable = !(role === 'identifier' && initialValue)

    const fieldProps = {
      ...this.props,
      autoComplete: 'off',
      className: 'u-m-0', // 0 margin
      disabled: disabled || !isEditable,
      error: hasError,
      fullwidth: true,
      label: t(`fields.${localeKey}.label`, {
        _: t(`legacy.fields.${localeKey}.label`, { _: name })
      }),
      onKeyUp: onKeyUp,
      placeholder: getFieldPlaceholder(
        this.props,
        t(`fields.${name}.placeholder`, { _: '' })
      ),
      side: !required && t('accountForm.fields.optional'),
      size: 'medium'
    }
    const passwordLabels = {
      hideLabel: t('accountForm.password.hide'),
      showLabel: t('accountForm.password.show')
    }
    switch (type) {
      case 'date':
        return (
          <Field
            {...fieldProps}
            placeholder={t(`fields.${name}.placeholder`, {
              _: t(`default.dateFormat`).toLowerCase()
            })}
          />
        )
      case 'dropdown':
        return <Field {...sanitizeSelectProps(fieldProps)} />
      case 'password':
        return (
          <Field
            {...fieldProps}
            // Usting the `new-password` value is the best way to avoid
            // autocomplete for password.
            // See https://stackoverflow.com/a/17721462/1135990
            autoComplete="new-password"
            secondaryLabels={passwordLabels}
          />
        )
      default:
        return <Field {...fieldProps} type="text" />
    }
  }
}

AccountField.propTypes = {
  hasError: PropTypes.bool,
  initialValue: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  label: PropTypes.string,
  name: PropTypes.string.isRequired,
  onKeyUp: PropTypes.func,
  role: PropTypes.string,
  type: PropTypes.oneOf(['date', 'dropdown', 'email', 'password', 'text']),
  t: PropTypes.func
}

// As SelectBox component from Cozy-UI, rendering dropdown type, is just giving
// us the full Option object, we just get its value to facilitate mapping
// with account
const parse = type => value => {
  return type === 'dropdown' ? value.value : value
}

export class AccountFields extends PureComponent {
  render() {
    const {
      container,
      disabled,
      hasError,
      initialValues,
      manifestFields,
      onKeyUp,
      t
    } = this.props

    // Ready to use named fields array
    const namedFields = Object.keys(manifestFields).map(fieldName => ({
      ...manifestFields[fieldName],
      name: fieldName
    }))

    return (
      <div>
        {namedFields.map((field, index) => (
          <FinalFormField
            key={index}
            name={field.name}
            parse={parse(field.type)}
          >
            {({ input }) => (
              <AccountField
                {...field}
                {...input}
                container={container}
                hasError={hasError}
                disabled={disabled}
                initialValue={
                  initialValues[field.name] ||
                  initialValues[getEncryptedFieldName(field.name)]
                }
                onKeyUp={onKeyUp}
                t={t}
              />
            )}
          </FinalFormField>
        ))}
      </div>
    )
  }
}

AccountFields.propTypes = {
  disabled: PropTypes.bool,
  hasError: PropTypes.bool,
  fillEncrypted: PropTypes.bool,
  manifestFields: PropTypes.object.isRequired,
  onKeyUp: PropTypes.func,
  t: PropTypes.func
}

export class AccountForm extends PureComponent {
  constructor(props, context) {
    super(props, context)
    const { konnector, lang } = props
    const { locales } = konnector
    if (locales && lang) {
      extend(locales[lang])
    }
  }

  /**
   * Indicates if the state of ReactFinalForm implies that data can be submitted
   * @param  {Object} formState See https://github.com/final-form/final-form#formstate
   * @return {Boolean}
   */
  isSubmittable({ dirty, error, initialValues, valid }) {
    const untouched = initialValues && !dirty
    return error || (valid && !untouched)
  }

  handleKeyUp(event, { dirty, form, initialValues, valid, values }) {
    if (
      event.code === 'Enter' &&
      this.isSubmittable({ dirty, initialValues, valid })
    ) {
      this.handleSubmit(values, form)
    }
  }

  handleSubmit(values, form) {
    const { onSubmit } = this.props
    // Reset form with new values to set back dirty to false
    form.reset(values)
    onSubmit(values)
  }

  validate = (fields, initialValues) => vals => {
    let errors = {}
    for (let name in fields)
      if (
        fields[name].required &&
        !vals[name] &&
        // Don't require value for empty encrypted fields with initial value
        !initialValues[getEncryptedFieldName(name)]
      )
        errors[name] = VALIDATION_ERROR_REQUIRED_FIELD
    return errors
  }

  render() {
    const { account, error, konnector, onSubmit, submitting, t } = this.props
    const { fields, oauth } = konnector

    if (oauth) return <OAuthForm initialValues={initialValues} oauth={oauth} />

    const sanitizedFields = manifest.sanitizeFields(fields)
    const defaultValues = manifest.defaultFieldsValues(sanitizedFields)
    const initialValues = account && account.auth
    const initialAndDefaultValues = { ...defaultValues, ...initialValues }

    let container = null

    return (
      <Form
        initialValues={initialAndDefaultValues}
        onSubmit={onSubmit}
        validate={this.validate(sanitizedFields, initialAndDefaultValues)}
        render={({ dirty, form, values, valid }) => (
          <div
            ref={element => {
              container = element
            }}
          >
            {error && (
              <AccountFormError error={error} konnector={konnector} t={t} />
            )}
            <AccountFields
              container={container}
              disabled={submitting}
              hasError={
                error &&
                error instanceof KonnectorJobError &&
                error.isLoginError()
              }
              initialValues={initialAndDefaultValues}
              manifestFields={sanitizedFields}
              onKeyUp={event =>
                this.handleKeyUp(event, {
                  dirty,
                  form,
                  initialValues,
                  valid,
                  values
                })
              }
              t={t}
            />
            <Button
              busy={submitting}
              className="u-mt-2 u-mb-1-half"
              disabled={
                submitting ||
                !this.isSubmittable({ dirty, error, initialValues, valid })
              }
              extension="full"
              label={t('accountForm.submit.label')}
              onClick={() => this.handleSubmit(values, form)}
            />
          </div>
        )}
      />
    )
  }
}

AccountForm.propTypes = {
  account: PropTypes.object,
  konnector: PropTypes.object.isRequired,
  error: PropTypes.object,
  submitting: PropTypes.bool
}

export default translate()(AccountForm)
