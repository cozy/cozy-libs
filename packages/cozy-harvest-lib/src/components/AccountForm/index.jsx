import React, { PureComponent } from 'react'
import { Form } from 'react-final-form'
import PropTypes from 'prop-types'

import { isMobile } from 'cozy-device-helper'
import Button from 'cozy-ui/transpiled/react/Button'
import { translate } from 'cozy-ui/transpiled/react/I18n'

import AccountFields from './AccountFields'
import TriggerErrorInfo from '../infos/TriggerErrorInfo'
import { getEncryptedFieldName } from '../../helpers/fields'
import { KonnectorJobError } from '../../helpers/konnectors'
import manifest from '../../helpers/manifest'
import withKonnectorLocales from '../hoc/withKonnectorLocales'

const VALIDATION_ERROR_REQUIRED_FIELD = 'VALIDATION_ERROR_REQUIRED_FIELD'

/**
 * AccountForm is reponsible of generating a form which will allow user to
 * edit an account for a given konnector.
 *
 * This component rely internally on a ReactFinalForm component.
 *
 * AccountForm must be passed `onSubmit` callback.
 * @type {PureComponent}
 * @see https://github.com/final-form/react-final-form#getting-started
 */
export class AccountForm extends PureComponent {
  constructor(props) {
    super(props)

    this.inputs = {}
    this.inputFocused = null

    this.inputRefByName = this.inputRefByName.bind(this)
    this.handleFocus = this.handleFocus.bind(this)
    this.handleKeyUp = this.handleKeyUp.bind(this)
    this.focusNext = this.focusNext.bind(this)
  }

  /**
   * Indicates if the state of ReactFinalForm implies that data can be submitted
   * @param  {Object} formState See https://github.com/final-form/final-form#formstate
   * @return {Boolean}
   */
  isSubmittable({ dirty, error, initialValues, valid }) {
    const modified = !initialValues || dirty
    // Here error is not a validation error but an instance of a
    // KonnectorJobError, so submitting again is possible
    return modified && (error || valid)
  }

  /**
   * Give focus to next input in the form.
   * Fallback for mobile devices as we are using a div element instead of a
   * form.
   * @return {Element} Focused element or null if no element has been focused
   */
  focusNext() {
    if (!this.inputs) return null

    const inputs = Object.values(this.inputs)

    const currentIndex = inputs.indexOf(this.inputFocused)

    let nextIndex = currentIndex + 1
    let nextInput = inputs[nextIndex]

    if (nextInput) {
      nextInput.focus()
    }

    return nextInput || null
  }

  /**
   * Capture input with focus and store the current focused input
   * @param  {Event} event Focus event
   */
  handleFocus(event) {
    if (Object.values(this.inputs).includes(event.target)) {
      this.inputFocused = event.target
    }
  }

  /**
   * Handle key up and check if `ENTER` key has been pressed. If so, submit the
   * form if all parmater are ok.
   * @param  {React.SyntheticEvent} event         Key events
   * @param  {Boolean} dirty         Indicates if form is dirty, i.e. if values
   * have changed
   * @param  {Object} form          The form object injected by ReactFinalForm.
   * @param  {Object} initialValues Initial data injected into AccountForm
   * @param  {Boolean} valid        Indicates if the form data is valid
   * @param  {Object} values        Actual form values data
   */
  handleKeyUp(event, { dirty, form, initialValues, valid, values }) {
    if (event.key === 'Enter') {
      const changedFocus = isMobile() && !!this.focusNext()
      if (
        !changedFocus &&
        this.isSubmittable({ dirty, initialValues, valid })
      ) {
        this.handleSubmit(values, form)
      }
    }
  }

  /**
   * Submit handler
   * @param  {Object} values        Actual form values data
   * @param  {Object} form          The form object injected by ReactFinalForm.
   */
  handleSubmit(values, form) {
    const { onSubmit } = this.props
    // Reset form with new values to set back dirty to false
    form.reset(values)
    onSubmit(values)
  }

  /**
   * Callback passed to `<AccountFields />` element. Called with a field name,
   * it return a callback wich take an input ref as argument. It then stores
   * the input ref into AccountForm's inner `inputs` property.
   * With this method we keep a list of inputs existing in the form.
   * The inputs inner property is indexed with names to avoid duplicates in case
   * of re-render of input children.
   * @param  {string} fieldName Field name
   * @return {function}         Callback which take an input ref as argument.
   */
  inputRefByName(fieldName) {
    return input => {
      this.inputs[fieldName] = input
    }
  }

  validate = (fields, initialValues) => vals => {
    let errors = {}
    for (let name in fields)
      if (
        fields[name].required &&
        !vals[name] &&
        // Don't require value for empty encrypted fields with initial value
        !initialValues[getEncryptedFieldName(name)]
      )
        errors[name] = VALIDATION_ERROR_REQUIRED_FIELD
    return errors
  }

  render() {
    const {
      account,
      error,
      konnector,
      onSubmit,
      showError,
      submitting,
      t
    } = this.props
    const { fields } = konnector

    const sanitizedFields = manifest.sanitizeFields(fields)
    const defaultValues = manifest.defaultFieldsValues(sanitizedFields)
    const initialValues = account && account.auth
    const initialAndDefaultValues = { ...defaultValues, ...initialValues }

    let container = null
    console.log('error AccountFrom', error)
    const isLoginError =
      error instanceof KonnectorJobError && error.isLoginError()

    return (
      // See https://github.com/final-form/react-final-form#getting-started
      <Form
        initialValues={initialAndDefaultValues}
        onSubmit={onSubmit}
        validate={this.validate(sanitizedFields, initialAndDefaultValues)}
        render={({ dirty, form, values, valid }) => (
          <div
            onKeyUp={event =>
              this.handleKeyUp(event, {
                dirty,
                form,
                initialValues,
                valid,
                values
              })
            }
            onFocusCapture={this.handleFocus}
            ref={element => {
              container = element
            }}
          >
            {error && showError && (
              <TriggerErrorInfo
                className="u-mb-1"
                error={error}
                konnector={konnector}
              />
            )}
            <AccountFields
              container={container}
              disabled={submitting}
              fields={sanitizedFields}
              hasError={error && isLoginError}
              initialValues={initialAndDefaultValues}
              inputRefByName={this.inputRefByName}
              t={t}
            />
            <Button
              busy={submitting}
              className="u-mt-2 u-mb-1-half"
              disabled={
                submitting ||
                !this.isSubmittable({ dirty, error, initialValues, valid })
              }
              extension="full"
              label={t('accountForm.submit.label')}
              onClick={() => this.handleSubmit(values, form)}
            />
          </div>
        )}
      />
    )
  }
}

AccountForm.propTypes = {
  /**
   * Account document. Used to get intial form values.
   * If no account is passed, AccountForm will use empty initial values.
   * @type {Object}
   */
  account: PropTypes.object,
  /**
   * Existing error
   * @type {Error,KonnectorJobError}
   */
  error: PropTypes.object,
  /**
   * Konnector document. AccountForm will check the `fields` object to compute
   * fields.
   * @type {Object}
   */
  konnector: PropTypes.object.isRequired,
  /**
   * Submit callback
   * @type {Function}
   */
  onSubmit: PropTypes.func.isRequired,
  /**
   * Indicates if the AccountForm has to show errors. Sometimes errors may be
   * displayed elsewhere. However, a KonnectorJobError corresponding to a login
   * error is always displayed.
   * @type {Boolean}
   */
  showError: PropTypes.bool,
  /**
   * Indicates if the form should be rendered as submitting data or busy.
   * Typically updated after an `onSubmit` call.
   * @type {Object}
   */
  submitting: PropTypes.bool,
  /**
   * Translation function
   */
  t: PropTypes.func
}

AccountForm.defaultProps = {
  showError: true
}

export default translate()(withKonnectorLocales(AccountForm))
