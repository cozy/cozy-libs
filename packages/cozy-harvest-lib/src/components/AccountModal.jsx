import React, { Component } from 'react'
import PropTypes from 'prop-types'

import get from 'lodash/get'
import { withMutations } from 'cozy-client'
import { withRouter } from 'react-router'
import Spinner from 'cozy-ui/transpiled/react/Spinner'
import { ModalContent } from 'cozy-ui/transpiled/react/Modal'
import Infos from 'cozy-ui/transpiled/react/Infos'
import Button from 'cozy-ui/transpiled/react/Button'
import { translate } from 'cozy-ui/transpiled/react/I18n'

import accountMutations from '../connections/accounts'
import triggersMutations from '../connections/triggers'
import * as triggersModel from '../helpers/triggers'
import KonnectorConfiguration from './KonnectorConfiguration/KonnectorConfiguration'
import AccountSelectBox from './AccountSelectBox/AccountSelectBox'
import KonnectorModalHeader from './KonnectorModalHeader'

/**
 * AccountModal take an accountId and a list of accounts containing their
 * respecting triggers and display the selected account and the accounts linked
 * to this konnector
 *
 * You have to pass an array of accounts containing their associated trigger:
 * ```
 * accounts: [
 *  {
 *    account: {
 *       _id: '',
 *    },
 *    trigger: {
 *      id: '',
 *      current_state: {
 *      }
 *    }
 *  }
 * ]
 * ```
 *
 * from this array and from the passed accountId, we can fetch the account.
 *
 */
class AccountModal extends Component {
  state = {
    trigger: null,
    account: null,
    fetching: true,
    error: false
  }
  componentDidMount() {
    this.loadSelectedAccountId()
  }

  componentDidUpdate(prevProps) {
    if (this.props.accountId !== prevProps.accountId)
      this.loadSelectedAccountId()
  }

  loadSelectedAccountId() {
    const { accountId, accounts } = this.props
    const matchingTrigger = get(
      accounts.find(account => account.account._id === accountId),
      'trigger'
    )
    if (matchingTrigger) this.fetchAccount(matchingTrigger)
  }

  async fetchAccount(trigger) {
    const { findAccount } = this.props
    this.setState({ fetching: true })

    try {
      const account = await findAccount(triggersModel.getAccountId(trigger))
      this.setState({
        account,
        trigger
      })
      return account
    } catch (error) {
      this.setState({
        error
      })
    } finally {
      this.setState({
        fetching: false
      })
    }
  }

  render() {
    const { konnector, onDismiss, history, accounts, t } = this.props
    const { trigger, account, fetching, error } = this.state
    return (
      <>
        <KonnectorModalHeader konnector={konnector}>
          {account && (
            <AccountSelectBox
              selectedAccount={account}
              accountsList={accounts}
              onChange={option => {
                history.push(`../${option.account._id}`)
              }}
              onCreate={() => {
                history.push(`../../new`)
              }}
            />
          )}
        </KonnectorModalHeader>
        <ModalContent>
          {error && (
            <Infos
              actionButton={
                <Button
                  theme="danger"
                  onClick={this.loadSelectedAccountId.bind(this)}
                  label={t('modal.konnector.error.button')}
                />
              }
              title={t('modal.konnector.error.title')}
              text={t('modal.konnector.error.description', error)}
              icon="warning"
              isImportant
            />
          )}
          {fetching && (
            <div className="u-ta-center">
              <Spinner size="xxlarge" />
            </div>
          )}
          {!error && !fetching && (
            <KonnectorConfiguration
              konnector={konnector}
              trigger={trigger}
              account={account}
              onAccountDeleted={onDismiss}
              addAccount={() => history.push('../new')}
            />
          )}
        </ModalContent>
      </>
    )
  }
}

AccountModal.propTypes = {
  konnector: PropTypes.object.isRequired,
  onDismiss: PropTypes.func,
  history: PropTypes.object.isRequired,
  accounts: PropTypes.array.isRequired,
  t: PropTypes.func.isRequired,
  findAccount: PropTypes.func.isRequired,
  accountId: PropTypes.string.isRequired
}
export default withMutations(accountMutations, triggersMutations)(
  withRouter(translate()(AccountModal))
)
