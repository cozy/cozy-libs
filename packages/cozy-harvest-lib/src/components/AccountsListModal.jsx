import React, { useContext } from 'react'
import PropTypes from 'prop-types'
import { ModalContent } from 'cozy-ui/transpiled/react/Modal'
import Stack from 'cozy-ui/transpiled/react/Stack'
import { translate } from 'cozy-ui/transpiled/react/I18n'
import AccountsList from './AccountsList/AccountsList'
import KonnectorIcon from './KonnectorIcon'
import { MountPointContext } from './MountPointContext'

const AccountsListModal = ({ konnector, accounts, t }) => {
  const { pushHistory } = useContext(MountPointContext)
  return (
    <>
      <ModalContent>
        <Stack className="u-mb-3">
          <div className="u-w-3 u-h-3 u-mh-auto">
            <KonnectorIcon konnector={konnector} />
          </div>
          <h3 className="u-title-h3 u-ta-center">
            {t('modal.accounts.title', { name: konnector.name })}
          </h3>
        </Stack>
        <AccountsList
          accounts={accounts}
          konnector={konnector}
          onPick={option => pushHistory(`/accounts/${option.account._id}`)}
          addAccount={() => pushHistory('/new')}
        />
      </ModalContent>
    </>
  )
}

AccountsListModal.propTypes = {
  konnector: PropTypes.object.isRequired,
  accounts: PropTypes.array.isRequired
}
export default translate()(AccountsListModal)
