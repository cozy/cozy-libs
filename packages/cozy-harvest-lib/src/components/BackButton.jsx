import React from 'react'
import Icon from 'cozy-ui/transpiled/react/Icon'
import cx from 'classnames'
import withLocales from './hoc/withLocales'

const BackButton = props => {
  const { className, t, ...rest } = props

  return (
    <button
      className={cx(
        'u-inline-flex u-flex-items-center u-bg-white u-bdw-0 u-ph-0 u-pv-1 u-c-pointer',
        className
      )}
      {...rest}
    >
      <Icon icon="left" className="u-mr-half" />
      {t('Revenir Ã  la liste')}
    </button>
  )
}

export default withLocales(BackButton)
