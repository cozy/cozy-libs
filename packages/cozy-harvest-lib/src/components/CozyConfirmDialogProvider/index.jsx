import React, { createContext, useContext, useState } from 'react'

import Button from 'cozy-ui/transpiled/react/Buttons'
import { ConfirmDialog } from 'cozy-ui/transpiled/react/CozyDialogs'

export const CozyConfirmDialogContext = createContext()

let currentDialogId = 0

export const useCozyConfirmDialog = () => {
  const context = useContext(CozyConfirmDialogContext)
  if (!context) {
    throw new Error(
      'Cannot use useCozyConfirmDialog without CozyConfirmDialogProvider. The component must a CozyConfirmDialogProvider above it.'
    )
  }
  return context
}

export const CozyConfirmDialogProvider = React.memo(props => {
  const [dialogs, setDialogs] = useState([])
  const { children } = props

  const onClose = dialogId => {
    dialogs.find(dialog => dialog.id === dialogId).callback()
    setDialogs(prevDialogs =>
      prevDialogs.filter(dialog => dialog.id !== dialogId)
    )
  }

  const context = {
    showDialog: ({ title, closeLabel, description }) => {
      return new Promise(resolve => {
        setDialogs(prevDialogs => {
          return [
            ...prevDialogs,
            {
              id: currentDialogId++,
              title,
              closeLabel,
              description,
              callback: resolve
            }
          ]
        })
      })
    }
  }

  return (
    <CozyConfirmDialogContext.Provider value={context}>
      {dialogs &&
        dialogs.map(dialog => {
          return (
            <ConfirmDialog
              open
              key={dialog.id}
              title={dialog.title}
              content={dialog.description}
              actions={
                <Button
                  label={dialog.closeLabel}
                  aria-label="Close dialog"
                  onClick={() => onClose(dialog.id)}
                />
              }
              onClose={() => onClose(dialog.id)}
            />
          )
        })}
      {children}
    </CozyConfirmDialogContext.Provider>
  )
})

CozyConfirmDialogProvider.displayName = 'CozyConfirmDialogProvider'
