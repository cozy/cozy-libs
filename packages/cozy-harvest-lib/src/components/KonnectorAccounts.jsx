import React from 'react'

import { withMutations } from 'cozy-client'
import { withRouter } from 'react-router'
import Spinner from 'cozy-ui/transpiled/react/Spinner'
import { ModalContent } from 'cozy-ui/transpiled/react/Modal'
import { translate } from 'cozy-ui/transpiled/react/I18n'
import Infos from 'cozy-ui/transpiled/react/Infos'
import Button from 'cozy-ui/transpiled/react/Button'

import accountMutations from '../connections/accounts'
import triggersMutations from '../connections/triggers'
import * as triggersModel from '../helpers/triggers'
import KonnectorModalHeader from './KonnectorModalHeader'

class KonnectorAccounts extends React.Component {
  state = {
    fetchingAccounts: false,
    error: null,
    accounts: []
  }

  /**
   * TODO We should not fetchAccounts and fetchAccount since we already have the informations
   * in the props. We kept this sytem for compatibility on the existing override.
   * Next tasks: remove these methods and rewrite the override
   */
  async componentDidMount() {
    await this.fetchAccounts()
  }

  componentDidUpdate(prevProps) {
    // After leaving the new account screen, we need to refetch accounts in case a new one was created
    if (
      this.props.location.pathname !== prevProps.location.pathname &&
      /\/new\/?$/.test(prevProps.location.pathname)
    ) {
      this.fetchAccounts()
    }
  }

  async fetchAccounts() {
    const triggers = this.props.konnector.triggers.data
    const { findAccount } = this.props
    this.setState({ fetchingAccounts: true })
    try {
      const accounts = (await Promise.all(
        triggers.map(async trigger => {
          return {
            account: await findAccount(triggersModel.getAccountId(trigger)),
            trigger
          }
        })
      )).filter(({ account }) => !!account)
      this.setState({ accounts, fetchingAccounts: false, error: null })
    } catch (error) {
      this.setState({ error, fetchingAccounts: false })
    }
  }
  renderError() {
    const { t } = this.props

    return (
      <ModalContent>
        <Infos
          actionButton={
            <Button
              theme="danger"
              onClick={this.fetchAccounts.bind(this)}
              label={t('modal.accounts.error.retry')}
            />
          }
          title={t('modal.accounts.error.title')}
          text={t('modal.accounts.error.description')}
          icon="warning"
          isImportant
        />
      </ModalContent>
    )
  }
  renderLoading() {
    return (
      <ModalContent>
        <div className="u-pv-2 u-ta-center">
          <Spinner size="xxlarge" />
        </div>
      </ModalContent>
    )
  }
  render() {
    // TODO error
    const { accounts, fetchingAccounts, error } = this.state
    const { konnector } = this.props
    return (
      <>
        <KonnectorModalHeader konnector={konnector} />
        {error && this.renderError()}
        {fetchingAccounts && this.renderLoading()}
        {!error && !fetchingAccounts && this.props.children(accounts)}
      </>
    )
  }
}

export default withMutations(accountMutations, triggersMutations)(
  withRouter(translate()(KonnectorAccounts))
)
