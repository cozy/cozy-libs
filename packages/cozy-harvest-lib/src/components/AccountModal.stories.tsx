import type { Meta, StoryObj } from '@storybook/react'
import { withRouter } from 'storybook-addon-react-router-v6'

import { AccountModal } from './AccountModal'
import ConfigurationTab from './KonnectorConfiguration/ConfigurationTab'
import { storybookClient } from '../../.storybook/StoryContainer'
import { AccountModalFixtures } from '../../.storybook/fixtures/AccountModal.fixtures'

const meta: Meta<typeof AccountModal> = {
  argTypes: {
    onDismiss: { action: 'onDismiss' },
    replaceHistory: { action: 'replaceHistory' }
  },
  decorators: [withRouter],
  args: {
    client: storybookClient,
    accountId: '1337',
    t: (key: string) => key,
    showAccountSelection: true,
    showNewAccountButton: true,
    accounts: [],
    initialActiveTab: 'configuration',
    Component: ConfigurationTab,
    location: {
      hash: '',
      key: '5j2q3d',
      pathname: '/connected/stub',
      search: '',
      state: null
    },
    accountsAndTriggers: AccountModalFixtures.accountsAndTriggers,
    konnector: AccountModalFixtures.konnector,
    breakpoints: {
      isMobile: false,
      isTablet: false,
      isDesktop: true
    }
  },
  component: AccountModal,
  parameters: {
    reactRouter: {
      routePath: '/connected/stub/accounts/1337/config'
    }
  }
}

export default meta

type Story = StoryObj<typeof AccountModal>

export const WithConfigurationTab: Story = {
  args: {
    Component: ConfigurationTab
  }
}
