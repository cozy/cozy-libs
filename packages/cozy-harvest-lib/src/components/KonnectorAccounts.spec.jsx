import React from 'react'
import { shallow } from 'enzyme'
import { KonnectorAccounts } from 'components/KonnectorAccounts'
import CozyClient from 'cozy-client'
import { fetchAccountsFromTriggers } from '../../src/connections/accounts'
import { fetchTrigger } from '../../src/connections/triggers'

jest.mock(
  'cozy-realtime',
  () =>
    function() {
      return {
        subscribe: jest.fn(),
        unsubscribe: jest.fn()
      }
    }
)
jest.mock('../../src/connections/accounts', () => ({
  fetchAccountsFromTriggers: jest.fn()
}))

jest.mock('../../src/connections/triggers', () => ({
  fetchTrigger: jest.fn()
}))

describe('KonnectorAccounts', () => {
  const client = new CozyClient({})
  it('should show a spinner', () => {
    const children = jest.fn()
    const component = shallow(
      <KonnectorAccounts
        konnector={{}}
        location={{}}
        client={client}
        t={jest.fn()}
      >
        {children}
      </KonnectorAccounts>
    )

    expect(component.getElement()).toMatchSnapshot()
    expect(children).not.toHaveBeenCalled()
  })

  it('should call children with the accounts and triggers list', async () => {
    const children = jest.fn()
    const trigger = { _id: 'abc', error: null }
    fetchAccountsFromTriggers.mockResolvedValueOnce([
      {
        account: { _type: 'io.cozy.accounts', _id: '123' },
        trigger
      }
    ])
    const component = shallow(
      <KonnectorAccounts
        konnector={{ triggers: { data: [trigger] } }}
        location={{}}
        client={client}
        t={jest.fn()}
      >
        {children}
      </KonnectorAccounts>
    )
    await component.instance().fetchAccounts()
    await component.update()

    expect(children).toHaveBeenCalledWith([
      {
        account: { _type: 'io.cozy.accounts', _id: '123' },
        trigger: { _id: 'abc', error: null }
      }
    ])

    fetchTrigger.mockResolvedValueOnce({
      _id: 'abc',
      error: 'LOGIN_FAILED'
    })
    await component.instance().handleJobUpdate({ trigger_id: 'abc' })
    await component.update()

    expect(children).toHaveBeenCalledWith([
      {
        account: { _type: 'io.cozy.accounts', _id: '123' },
        trigger: { _id: 'abc', error: 'LOGIN_FAILED' }
      }
    ])
  })
})
