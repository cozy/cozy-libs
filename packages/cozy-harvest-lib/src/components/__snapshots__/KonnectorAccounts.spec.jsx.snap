// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KonnectorAccounts should show a spinner 1`] = `
<React.Fragment>
  <KonnectorModalHeader
    konnector={Object {}}
  />
  <ForwardRef(WithStyles)
    className="u-pb-2"
  >
    <div
      className="u-pv-2 u-ta-center"
    >
      <ForwardRef
        size="xxlarge"
      />
    </div>
  </ForwardRef(WithStyles)>
</React.Fragment>
`;
